### PR TITLE
Add smoke test for PVB user journey

### PR DESCRIPTION
**Pull request by dan-palmer**
Adds a smoke test flow to the application - found under `smoke_test/`.

Adds a `PrisonMailer` check for smoke tests - if its a smoke test avoid sending stuff to the actual prison!

##### Smoke Test Steps:
  - [x] Starts the form
  - [x] Fills in prisoner details page
  - [x] Fills in visitor details page for one visitor
  - [x] Selects three available days on the slot page
  - [x] Submits the booking on the check your request page
  - [x] Checks that the user received a booking receipt email
  - [x] Checks that the prison received a booking request email
  - [x] Visits the process booking page and confirms the booking
  - [x] Checks that the prison received a copy of the confirmation email
  - [x] Checks that the visitor received a confirmation email
  - [x] Checks the status of the visit and cancels it
  - [x] Checks that the prison receives a booking cancellation email

#### Improvements
- [x] Handle sending Prison mails through Sendgrid(see: #297)
- [x] Use unique gmail addresses (this one cool '+' trick)
- [x] Auto run the smoke test
- [x] Separate concerns a wee bit
- [x] Start the process form the '/prisoner' page to avoid routing issues with production

##### Required environment variables for the rails app:
```
SMOKE_TEST_EMAIL_LOCAL_PART
SMOKE_TEST_EMAIL_DOMAIN
```

##### Required environment variables for the smoke test:
```
SMOKE_TEST_EMAIL_LOCAL_PART
SMOKE_TEST_EMAIL_DOMAIN
SMOKE_TEST_EMAIL_PASSWORD
SMOKE_TEST_APP_HOST
```

### CODE REVIEW

File: smoke_test/smoke_test.rb [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  end
+
+  def state
+    @state ||= State.new
+  end
+
+  def start_form
+    visit '/'
+    click_link 'Start now'
+  end
```
Comment (SteveMarshall) [inline smoke_test/smoke_test.rb]: Probably don't wanna do this step this way, do we, given the number of emails we sent to Cardiff… Maybe just go straight to `/prisoner` on `SMOKE_TEST_APP_HOST`?

File: smoke_test/smoke_test.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+Capybara.run_server = false
+Capybara.current_driver = :poltergeist
+Capybara.app_host = ENV.fetch('SMOKE_TEST_APP_HOST')
+
+module SmokeTest
+  extend Capybara::DSL
+
+  SMOKE_TEST_EMAIL_LOCAL_PART = ENV.fetch 'SMOKE_TEST_EMAIL_LOCAL_PART'
+  SMOKE_TEST_EMAIL_DOMAIN = ENV.fetch 'SMOKE_TEST_EMAIL_DOMAIN'
```
Comment (threedaymonk) [inline smoke_test/smoke_test.rb]: These lines can be deleted

File: smoke_test/mail_box.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
+require 'mail'
+require 'net/imap'
+
+module SmokeTest
+  module MailBox
+    SMOKE_TEST_EMAIL_PASSWORD = ENV.fetch 'SMOKE_TEST_EMAIL_PASSWORD'
+    SMOKE_TEST_EMAIL_HOST = ENV.fetch 'SMOKE_TEST_EMAIL_HOST'
```
Comment (threedaymonk) [inline smoke_test/mail_box.rb]: These are already available in the outer scope thanks to `smoke_test.rb`

Comment (dan-palmer) [inline smoke_test/mail_box.rb]: Ahhh I'm such a dufus.. so upon closer inspection these are different constants to what `SmokeTest` offers :sob:

Comment (threedaymonk) [inline smoke_test/mail_box.rb]: Oops, so they are. The fact that they deceive the eye with their similarity is possibly a warning flag.

File: config/environments/development.rb [hunk 4]
```diff
@@ -30,4 +30,7 @@
   config.action_mailer.default_url_options = { host: "localhost", protocol: "http", port: "3000" }
 
   config.action_mailer.preview_path = "#{Rails.root}/test/mailers/previews"
+
+  config.smoke_test_email_local_part = ENV.fetch('SMOKE_TEST_EMAIL_LOCAL_PART')
+  config.smoke_test_email_domain = ENV.fetch('SMOKE_TEST_EMAIL_DOMAIN')
```
Comment (threedaymonk) [inline config/environments/development.rb]: This is going to make running `rails s` a hassle.

**Comment by threedaymonk** [comment]
Looks fine to me except for the `ENV.fetch` in `config/environments/development.rb`. I'm find with `fetch`, but I think we should provide some sensible defaults either via a development env file or in `fetch` itself.

**Comment by dan-palmer** [comment]
Opted to use `ENV[]` so that we don't have to fiddle with local environment variables or the PVB deploy scripts. Lets handle this better in a totally separate issue.

**Comment by SteveMarshall** [comment]
@dan-palmer Have you got a Jenkins job set up to run this yet?

**Pull request merged**